hail/python/hail/ir/renderer.py

```diff
@@ -134,3 +135,563 @@ def __call__(self, x: 'Renderable'):
+    #
+    # Returns a Dict summarizing of all lets to be inserted. For each 'node'
+    # which will have lets inserted immediately above it, maps 'id(node)' to a
+    # 'BindingSite' recording the depth of 'node' and a Dict 'lifted_lets',
+    # where for each descendant 'x' which will be bound above 'node',
+    # 'lifted_lets' maps 'id(x)' to the unique id 'x' will be bound to.
+    def __call__(self, root: 'ir.BaseIR') -> Dict[int, BindingSite]:
+        root_frame = self.StackFrame(0, 0, False, ({}, {}, {}), root)
+        stack = [root_frame]
+        binding_sites = {}
+
+        while True:
+            frame = stack[-1]
+            node = frame.node
+            frame.child_idx += 1
+            child_idx = frame.child_idx
+
+            if child_idx >= len(node.children):
+                # mark node as visited at potential let insertion site
+                if not node.is_effectful():
+                    bind_depth = frame.bind_depth()
+                    if bind_depth < frame.min_value_binding_depth:
+                        if frame.scan_scope:
+                            stack[bind_depth].scan_visited.add(id(node))
+                        else:
+                            stack[bind_depth].agg_visited.add(id(node))
+                    else:
+                        stack[bind_depth].visited.add(id(node))
+
+                stack.pop()
+
+                # if any lets being inserted here, add node to registry of
+                # binding sites
+                if frame.has_lifted_lets():
+                    binding_sites[id(node)] = frame.make_binding_site(len(stack))
+
+                if not stack:
+                    break
+                stack[-1].update_free_vars(frame)
+                continue
+
+            child = node.children[child_idx]
+
+            if self.renderer.stop_at_jir and hasattr(child, '_jir'):
+                self.renderer._add_jir(child)
+                continue
+
+            child_frame = frame.make_child_frame(len(stack))
+
+            if isinstance(child, ir.IR):
+                for i in reversed(range(child_frame.min_binding_depth, len(stack))):
+                    cur = stack[i]
+                    if i >= child_frame.min_value_binding_depth:
+                        if id(child) in cur.visited:
+                            seen_in_scope = i
+                            lets = cur.lifted_lets
+                            break
+                    else:
+                        if id(child) in cur.agg_visited:
+                            seen_in_scope = i
+                            lets = cur.agg_lifted_lets
+                            break
+                        if id(child) in cur.scan_visited:
+                            seen_in_scope = i
+                            lets = cur.scan_lifted_lets
+                            break
+                else:
+                    lets = None
+
+                if lets is not None:
```

Comment:
it seems possible for `lets` to not be defined here, is that true? 

If we hit the `else` on 221, and neither of the `if`s on 222 or 226, `lets` will be a name error here.

Reply:
The `lets = None` is the else clause of the for loop on 214, which is executed if we get through the for loop without triggering a break. All of the breaks follow setting `lets`, and if none of those happen, we set `lets = None` in the else clause.